tests/test_association.py

```diff
@@ -0,0 +1,357 @@
+            metadata,
+            'association',
+            'association.fracUpdatedDIAObjects')
+        self.assertIsInstance(meas, Measurement)
+        self.assertEqual(
+            meas.metric_name,
+            lsst.verify.Name(
+                metric='association.fracUpdatedDIAObjects'))
+        self.assertEqual(meas.quantity, 7 * u.count)
+
+        meas = measure_fraction_updated_dia_objects(
+            metadata,
+            'association',
+            'association.fracUpdatedDIAObjects')
+        self.assertIsInstance(meas, Measurement)
+        self.assertEqual(
+            meas.metric_name,
+            lsst.verify.Name(metric='association.fracUpdatedDIAObjects'))
+        self.assertEqual(meas.quantity, 5 / (5 + 7) * u.dimensionless_unscaled)
+
+
+    def test_valid_from_butler(self):
+        """ Test the association measurements that require a butler.
+        """
+        meas = measure_number_sci_sources(
+            self.butler,
+            dataId_dict=dataId_dict,
+            metric_name='ip_diffim.numSciSrc')
+        self.assertIsInstance(meas, Measurement)
+        self.assertEqual(
+            meas.metric_name,
+            lsst.verify.Name(metric='ip_diffim.numSciSrc'))
+        self.assertEqual(meas.quantity, 10 * u.count)
+
+        meas = measure_fraction_dia_sources_to_sci_sources(
+            self.butler,
+            dataId_dict=dataId_dict,
+            metric_name='ip_diffim.fracDiaSrcToSciSrc')
+        self.assertIsInstance(meas, Measurement)
+        self.assertEqual(
+            meas.metric_name,
+            lsst.verify.Name(metric='ip_diffim.fracDiaSrcToSciSrc'))
+        # We put in half the number of DIASources as detected sources.
+        self.assertEqual(meas.quantity, 0.5 * u.dimensionless_unscaled)
+
+
+    def test_valid_from_sqlite(self):
+        conn = sqlite3.connect(self.db_file)
+        cursor = conn.cursor()
+
+        meas = measure_total_unassociated_dia_objects(
+            cursor,
+            metric_name='association.numTotalUnassociatedDiaObjects')
+        self.assertIsInstance(meas, Measurement)
+        self.assertEqual(
+            meas.metric_name,
+            lsst.verify.Name(
+                metric='association.numTotalUnassociatedDiaObjects'))
+        self.assertEqual(meas.quantity, 5 * u.count)
+
+    def test_no_butler_data(self):
+        """ Test attempting to create a measurement with data that the butler
+        does not contain.
+        """
+
+        with self.assertRaises(dafPersist.NoResults):
+            measure_number_sci_sources(
+                self.butler,
+                dataId_dict={'visit': 1000, 'filter': 'r'},
+                metric_name='ip_diffim.fracDiaSrcToSciSrc')
+
+        with self.assertRaises(dafPersist.NoResults):
+            measure_fraction_dia_sources_to_sci_sources(
+                self.butler,
+                dataId_dict={'visit': 1000, 'filter': 'r'},
+                metric_name='ip_diffim.fracDiaSrcToSciSrc')
```

Comment:
You might also check that you get an error if you supply the correct visit number, but a missing filter.